### ISSUE TITLE
pyright-langserver reporting wrong Content-Length

### ISSUE THREAD

**Issue by neelkarma** [microsoft/pyright#1576]
Note: if you are reporting a wrong signature of a function or a class in the standard library, then the typeshed tracker is better suited for this report: https://github.com/python/typeshed/issues.

**Describe the bug**
When pyright-langserver reports an error, it's `Content-Length` header is wrong, which breaks linting (the editor doesn't report any errors even when there are).

**To Reproduce**
Write any error-producing code in any LSP-enabled text editor, and enable logging. The `Content-Length` of `textDocument/publishDiagnostics` messages will be incorrect.

(I'm not sure how LSP servers work, so please dont be harsh lol)

**Expected behavior**
Linting would work as expected.

**Screenshots or Code**
Write this in your preferred LSP-enabled text editor:
```py
def printInt(integer: int) -> None:
    print(integer)

printInt("hello") # Should cause error
```
For me, pyright-langserver sent this (among other unrelated things):
```
Content-Length: 602^M^@^M^@{"jsonrpc":"2.0","method":"textDocument/publishDiagnostics","params":{"uri":"file:///c%3A/Users/nsdjs/Documents/Programming/Python/lintertest/lintertest.py","diagnostics":[{"range":{"start":{"line":3,"character":9},"end":{"line":3,"character":14}},"message":"Argument of type \"Literal[''432'']\" cannot be assigned to parameter \"integer\" of type \"int\" in function \"printInt\"\n  \"Literal[''432'']\" is incompatible with \"int\"","severity":1,"code":"reportGeneralTypeIssues","source":"Pyright","codeDescription":{"href":"https://github.com/microsoft/pyright/blob/master/docs/configuration.md"}}]}}
```
If you check the length of the JSON object, it shows 604 instead of the reported 602:
```py
>>> len(br"""{"jsonrpc":"2.0","method":"textDocument/publishDiagnostics","params":{"uri":"file:///c%3A/Users/nsdjs/Documents/Programming/Python/lintertest/lintertest.py","diagnostics":[{"range":{"start":{"line":3,"character":9},"end":{"line":3,"character":14}},"message":"Argument of type \"Literal[''432'']\" cannot be assigned to parameter \"integer\" of type \"int\" in function \"printInt\"\n  \"Literal[''432'']\" is incompatible with \"int\"","severity":1,"code":"reportGeneralTypeIssues","source":"Pyright","codeDescription":{"href":"https://github.com/microsoft/pyright/blob/master/docs/configuration.md"}}]}}""")
604
```

**VS Code extension or command-line**
Are you running pyright as a VS Code extension or a command-line tool? Which version? You can find the version of the VS Code extension by clicking on the Pyright icon in the extensions panel.

Nope - using NeoVim 0.4.4 with ALE (which uses pyright-langserver).

**Additional context**
Here's the issue this issue originated from: dense-analysis/ale#3596

**Comment by erictraut** [comment]
Pyright doesn't create these message responses directly. It uses the vscode-languageserver library to do so. If this is a bug, it would need to be fixed in that library.

@jakebailey , thought?

**Comment by jakebailey** [comment]
The URI escaping is done well before the Content-Length is calculated (and is fixed afterward), so I think that's a red herring. You could name your file as something with another colon it it to see if the length increases to tell for sure, though.

I'm not personally aware of any issue like this; usually if the server sends something with the wrong content length, things catastrophically fail, whereas the linked issue makes it seem like it's only diagnostics which fail (pyright sends messages to the client for other reasons too). I saw this previously in C# where the JSON-RPC lib was adding the length of a UTF-8 BOM (yuck), but that would be 3 bytes different.

**Comment by erictraut** [comment]
@dbaeumer , any suggestions on this one? I suspect it might be a bug in vscode-languageserver's calculation of the JSON RPC content length.

**Comment by erictraut** [comment]
I'm going to transfer this issue to the vscode-languageserver.

**Comment by erictraut** [comment]
Wasn't able to transfer, so I created a new issue that links to this one.

**Comment by dbaeumer** [comment]
@erictraut Any steps on how to reproduce this. 

In the LSP libraries the content length is computed generically for all sorts of notifications and requests / responses using the Byte Buffer (see here: https://github.com/microsoft/vscode-languageserver-node/blob/main/jsonrpc/src/common/messageWriter.ts#L135). I don't try to say that there isn't a bug but without steps very hard to tell what is going on.

**Comment by erictraut** [comment]
I don't have a repro beyond what @neelkarma reports above.

The message that he pasted comes from the `_Connection.sendDiagnostics` callback in vscode_languageserver. It's called [here](https://github.com/microsoft/pyright/blob/4fa1acf704affd65e15f71f49bdb508f6493f58d/packages/pyright-internal/src/languageServerBase.ts#L1039). I don't see anything unusual about the way we're calling it. 

I did just notice that there is an escape sequence ("%3A") in the message pasted above. That might explain the discrepancy in length, since "%3A" is exactly two bytes longer than ":". Is the Content-Length computed before or after escaping?

**Comment by dbaeumer** [comment]
Actually I took the strings provided in the description and wrote a little script that uses the same code as the library to determine the content length and the result is 587.

```js
const s = '{"jsonrpc":"2.0","method":"textDocument/publishDiagnostics","params":{"uri":"file:///c%3A/Users/nsdjs/Documents/Programming/Python/lintertest/lintertest.py","diagnostics":[{"range":{"start":{"line":3,"character":9},"end":{"line":3,"character":14}},"message":"Argument of type \"Literal[\'432\']\" cannot be assigned to parameter \"integer\" of type \"int\" in function \"printInt\"\n  \"Literal[\'432\']\" is incompatible with \"int\"","severity":1,"code":"reportGeneralTypeIssues","source":"Pyright","codeDescription":{"href":"https://github.com/microsoft/pyright/blob/master/docs/configuration.md"}}]}}';

const buffer = Buffer.from(s, 'utf8');
console.log(`Content Length: ${buffer.byteLength}`);
```

So without further steps it is hard to tell what failed. 

@neelkarma can you please provide steps on how to reproduce this in a running example. Appreciated would be a GitHub repository that I can clone to ensure we are both looking at the same setup.

**Comment by erictraut** [comment]
@dbaeumer, if you paste the text as raw (unencoded) text, its length is 604 bytes.

@neelkarma, since you are seeing different behaviors with different clients, that's a pretty strong indication that it's not a problem with the language server. It sounds like it may be a problem with the NeoVim LSP layer. Please reach out to the maintainers of that components. Let us know if we can help further, but for now, I'm going to close this bug.

**Comment by dbaeumer** [comment]
@erictraut correct. My fault. The correct string is

```
const s = '{"jsonrpc":"2.0","method":"textDocument/publishDiagnostics","params":{"uri":"file:///c%3A/Users/nsdjs/Documents/Programming/Python/lintertest/lintertest.py","diagnostics":[{"range":{"start":{"line":3,"character":9},"end":{"line":3,"character":14}},"message":"Argument of type \\"Literal[\\\'432\\\']\\" cannot be assigned to parameter \\"integer\\" of type \\"int\\" in function \\"printInt\\"\\n  \\"Literal[\\\'432\\\']\\" is incompatible with \\"int\\"","severity":1,"code":"reportGeneralTypeIssues","source":"Pyright","codeDescription":{"href":"https://github.com/microsoft/pyright/blob/master/docs/configuration.md"}}]}}';

```

**Comment by dbaeumer** [comment]
I created a better example since the jsonrpc layer in LSP receives JSON objects not strings and does the JSON -> string internally. However still not sure how the original JSON object looked like. From the string I inferred this. In case we need to follow up on this.

```
const json = {
	"jsonrpc": "2.0",
	"method": "textDocument/publishDiagnostics",
	"params": {
		"uri": "file:///c%3A/Users/nsdjs/Documents/Programming/Python/lintertest/lintertest.py",
		"diagnostics": [
			{
				"range": {
					"start": {
						"line": 3,
						"character": 9
					},
					"end": {
						"line": 3,
						"character": 14
					}
				},
				"message": "Argument of type \"Literal['432']\" cannot be assigned to parameter \"integer\" of type \"int\" in function \"printInt\"\n  \"Literal['432']\" is incompatible with \"int\"",
				"severity": 1,
				"code": "reportGeneralTypeIssues",
				"source": "Pyright",
				"codeDescription": {
					"href":"https://github.com/microsoft/pyright/blob/master/docs/configuration.md"
				}
			}
		]
	}
};
const buffer = Buffer.from(JSON.stringify(json, undefined, 0), 'utf8');
console.log(`Content Length: ${buffer.byteLength}`);
```
